pkg/deploy/deploy.go

```diff
@@ -75,76 +77,111 @@ func DownloadManifests(ctx context.Context, componentName string, manifestConfig
+func unpackTarFromReader(reader io.Reader, basePath, componentName, contextDir string) error {
+	tarReader := tar.NewReader(reader)
 
-	// Extract the contents of the TAR archive to the current directory
 	for {
 		header, err := tarReader.Next()
 		if errors.Is(err, io.EOF) {
 			break
 		}
+		if err != nil {
+			return fmt.Errorf("error reading tar header: %w", err)
+		}
+
+		targetPath, err := resolveTargetPath(header.Name, basePath, componentName, contextDir)
+		if errors.Is(err, errPathIrrelevant) {
+			continue
+		}
 		if err != nil {
 			return err
 		}
 
-		componentFiles := strings.Split(header.Name, "/")
-		componentFileName := header.Name
-		componentManifestPath := componentFiles[0] + "/" + manifestConfig.ContextDir
+		err = processTarHeader(header, tarReader, targetPath)
+		if err != nil {
+			return err
+		}
+	}
 
-		if strings.Contains(componentFileName, componentManifestPath) {
-			// Get manifest path relative to repo
-			// e.g. of repo/a/b/manifests/base --> base/
-			componentFoldersList := strings.Split(componentFileName, "/")
-			componentFileRelativePathFound := strings.Join(componentFoldersList[len(strings.Split(componentManifestPath, "/")):], "/")
+	return nil
+}
 
-			if header.Typeflag == tar.TypeDir {
-				err = os.MkdirAll(DefaultManifestPath+"/"+componentName+"/"+componentFileRelativePathFound, mode)
-				if err != nil {
-					return fmt.Errorf("error creating directory:%w", err)
-				}
+// resolveTargetPath computes the target file path based on the tar header and context directory.
+func resolveTargetPath(headerName, basePath, contextDir, componentName string) (string, error) {
+	componentFiles := strings.Split(headerName, "/")
+	componentManifestPath := filepath.Join(componentFiles[0], contextDir)
 
-				continue
-			}
+	if !strings.Contains(headerName, componentManifestPath) {
+		return "", errPathIrrelevant
+	}
 
-			if header.Typeflag == tar.TypeReg {
-				file, err := os.Create(DefaultManifestPath + "/" + componentName + "/" + componentFileRelativePathFound)
-				if err != nil {
-					return fmt.Errorf("error creating file: %w", err)
-				}
-
-				defer file.Close()
-
-				for {
-					_, err := io.CopyN(file, tarReader, 1024)
-					if err != nil {
-						if errors.Is(err, io.EOF) {
-							break
-						}
-						return fmt.Errorf("error downloading file contents: %w", err)
-					}
-				}
-
-				continue
-			}
-		}
+	componentFoldersList := strings.Split(headerName, "/")
+	if len(componentFoldersList) < len(strings.Split(componentManifestPath, "/")) {
+		return "", errPathResolutionFailed // Path resolution failed
 	}
 
-	return err
+	relativePath := strings.Join(componentFoldersList[len(strings.Split(componentManifestPath, "/")):], "/")
+
+	return filepath.Join(basePath, componentName, relativePath), nil
+}
+
+// processTarHeader processes a TAR header, creating files or directories as needed.
+func processTarHeader(header *tar.Header, tarReader *tar.Reader, targetPath string) error {
```

Comment:
```suggestion
func extractFileOrDirectory(header *tar.Header, tarReader *tar.Reader, targetPath string) error {
```

I wonder if this might be a better name, since that's the effect that it has; whereas looking at the tar header seems to be kind of an internal detail?

Reply:
+1

Reply:
@grdryn,  sorry for the delay—I missed the notification. I agree and have applied the commit.